### DISCUSSION TITLE
How would I implement an event registry?

### DISCUSSION THREAD

**Discussion by J-Cake** [mlua-rs/rlua#262] [Q&A] [answered]
Hi, I'm using rlua for a plugin system where I need to expose event handler functions. I wish to achieve this by capturing a set of predefined globals. If they're functions, call them when ever the corresponding event occurs. Take the following lua code:

```lua

function on_event_1(event)
  print("Event 1")
end

function not_an_event_handler()
  print("Not an event handler")
end

```

I'd capture the handlers like so:

```rust
lua.context(|ctx| {
  // ...
  ctx.load(/*code*/).exec();

  if let Ok(handler) = ctx.globals().get::<_, rlua::Function>("on_event_1") { 
    self.on_event_1 = handler;
  }

});
```

But I get issues because the `ctx` object is leaked from the closure. Makes sense, the plugin outlives the context. 

My question would be about what the best way to achieve this would be?

**Comment by jugglerchris** [comment, marked as answer]
Hi,
I think what you want is probably [Context::create_registry_value](https://docs.rs/rlua/latest/rlua/struct.Context.html#method.create_registry_value) to create a reference you can store outside of the context, along with [Context::registry_value](https://docs.rs/rlua/latest/rlua/struct.Context.html#method.registry_value) to retrieve it when needed.

Chris

**Comment by J-Cake** [comment]
Hi Chris, 

Thanks for your quick reply. I ended up using your suggestion and it's working well. (my project isn't huge yet so it was an easy implementation)

Thanks
Jake

**Comment by jugglerchris** [comment]
Great, glad to hear it!
